iwyu.cc

```diff
@@ -1650,13 +1676,34 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
       return;
 
     // Figure out the best location to attribute uses inside macros.
-    if (IsInMacro(used_loc))
-      used_loc = GetUseLocationForMacroExpansion(used_loc, decl);
     const FileEntry* used_in = GetFileEntry(used_loc);
+    const NamedDecl* fwd_decl = NULL;
+    bool isMacroDefinedAndUsedInSameFile = false;
+    if (IsInMacro(used_loc)) {
+      isMacroDefinedAndUsedInSameFile =
+          (GetFileEntry(used_loc) ==
+           GetFileEntry(GetInstantiationLoc(used_loc)));
+
+      const NamedDecl* fwd_decl = GetForwardDeclareInSameFile(decl, used_loc);
+      if (fwd_decl) {
+        // Make sure we keep the heuristic forward decl.
+        preprocessor_info().FileInfoFor(used_in)->ReportForwardDeclareUse(
+            used_loc, decl, IsNodeInsideCXXMethodBody(current_ast_node()),
+            comment);
+
+        // Attribute use to caller
+        used_loc = GetInstantiationLoc(used_loc);
+      } else {
+        used_loc = GetUseLocationForMacroExpansion(used_loc, decl);
+      }
+    }
 
-    preprocessor_info().FileInfoFor(used_in)->ReportForwardDeclareUse(
-        used_loc, decl, IsNodeInsideCXXMethodBody(current_ast_node()),
-        comment);
+    used_in = GetFileEntry(used_loc);
+    if (isMacroDefinedAndUsedInSameFile || !fwd_decl) {
```

Comment:
I'm working through this now, and I can't make any sense of it myself. I'll see if I can make it clearer.